### PR TITLE
Fix camera flip on mobile

### PR DESCRIPTION
**Pull request by ROBYER1**
-Fix flipped front camera on mobile
-Allow default resolution of camera to be set to 1280x720 for devices that have high resolution cameras (prevents crash or slow camera behaviour on mobile for high resolution camera devices)

### CODE REVIEW

**Comment by ROBYER1** [comment]
I wasn't sure if flipping the screen's local scale was the best way to do this but it seemed to have the least ramifications

**Comment by homuler** [comment]
Thanks for the PR!
I think there are two different revisions (i.e. camera flip and resolution) in this PR, and it would be great if they could be separated into separate PRs.

> - Allow default resolution of camera to be set to 1280x720 for devices that have high resolution cameras (prevents crash or slow camera behaviour on mobile for high resolution camera devices)

I agree that it is better to set a default value for resolution, but I think it should not be fixed, because the camera may not support the resolution.
How about choosing a resolution that is close to the determined value among `availableResolutions` ?
For example, we can choose the one whose longer side is closest to 1280 pixels.

**Comment by ROBYER1** [comment]
Thanks, I'll sort this next week - I'm fairly new to making PR's so apologies for doubling up on this one!

**Comment by ROBYER1** [comment]
Updated this PR with just the fix for flipped camera on mobile - particularly helpful for the pose graph as it needs to have accurate camera flip. Keen to hear your thoughts on this way of doing it as I wasn't sure if we just wanted to do more CPU intensive flipping of the webcam texture or just inverting the local scale of the screen which I have done.

File: Assets/Mediapipe/Samples/Common/Scripts/Solution.cs [hunk 1]
```diff
@@ -72,7 +72,33 @@ public virtual void Stop()
     protected static void SetupScreen(RawImage screen, ImageSource imageSource)
     {
       screen.rectTransform.sizeDelta = new Vector2(imageSource.textureWidth, imageSource.textureHeight);
-      screen.rectTransform.localEulerAngles = imageSource.rotation.Reverse().GetEulerAngles();
+      screen.rectTransform.localEulerAngles = imageSource.rotation.Reverse().GetEulerAngles();
+
+      //Check if the screen is flipped on mobile and flip it if it is the front camera
```
Comment (homuler) [inline Assets/Mediapipe/Samples/Common/Scripts/Solution.cs:77]: Always insert a space character after `//`.
```suggestion
      // Check if the screen is flipped on mobile and flip it if it is the front camera
```

The same goes for other parts.

**Review by homuler — CHANGES_REQUESTED** [review]

File: Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs [hunk 2]
```diff
@@ -69,7 +69,9 @@ public enum SourceType
     ///   If <see cref="type" /> does not support frame rate, it returns zero.
     /// </remarks>
     public virtual double frameRate => resolution.frameRate;
-    public float focalLengthPx { get; } = 2.0f; // TODO: calculate at runtime
+    public float focalLengthPx { get; } = 2.0f; // TODO: calculate at runtime
```
[the remark this answers is not in the document]
Comment (ROBYER1) [reply Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs:72]: This caught me off guard when forking this repo and making a PR with different line endings - in the main repo I would change this setting - Project Settings - Editor - Line Endings for New Scripts > set to Unix if those are the line endings you want, I have done the same on my repo 
image

File: Assets/Mediapipe/Samples/Common/Scripts/Solution.cs [hunk 3]
```diff
@@ -72,7 +72,33 @@ public virtual void Stop()
...
-      screen.rectTransform.localEulerAngles = imageSource.rotation.Reverse().GetEulerAngles();
+      screen.rectTransform.localEulerAngles = imageSource.rotation.Reverse().GetEulerAngles();
+
+      //Check if the screen is flipped on mobile and flip it if it is the front camera
+      if (imageSource.isFrontCamera == true)
+      {
+        if (imageSource.isHorizontallyFlipped == false)
+        {
+          //Check if the screen is rotated first
+          if (screen.gameObject.transform.rotation.z == 0)
```
[the remark this answers is not in the document]
Comment (ROBYER1) [reply Assets/Mediapipe/Samples/Common/Scripts/Solution.cs]: This check is needed on android, without it, the camera on launch of the app is rotated by 90 degrees - I actually couldn't figure out why it was happening. You will see the same behaviour if you remove the 'if (screen.gameObject.transform.rotation.z == 0)' checks around this or if you debug out the rotation of the screen. 

Is there somewhere else in your code that the screen gets rotated sometimes on mobile builds? I could add a check if z is 0/180/360 I guess?

File: Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (homuler) [reply Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs:72]: This is not a new script and now you've changed whole EOL to `CRLF`...
https://github.com/homuler/MediaPipeUnityPlugin/pull/344/files#diff-e8e16e3f1068f4c3fd7445b2f85b9391738281c1d5d84a2d32e7eebbcb68b617

File: Assets/Mediapipe/Samples/Common/Scripts/ImageSource/WebCamSource.cs [hunk 4]
```diff
@@ -172,10 +172,11 @@ public override void SelectSource(int sourceId)
       {
         throw new ArgumentException($"Invalid source ID: {sourceId}");
       }
-
```
Comment (homuler) [inline Assets/Mediapipe/Samples/Common/Scripts/ImageSource/WebCamSource.cs:175]: [nits] This change is unnecessary.

**Review by homuler — CHANGES_REQUESTED** [review]

File: Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ROBYER1) [reply Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs:72]: I'll sort this out tomorrow - my first PR on a non CRLF repo so I was equally baffled when I thought I sorted line endings to be Unix LF only in Unity + Visual Studio Community 2022. I'll sort this out

File: Assets/Mediapipe/Samples/Common/Scripts/ImageSource/WebCamSource.cs (hunk 4 above)
Comment (ROBYER1) [reply Assets/Mediapipe/Samples/Common/Scripts/ImageSource/WebCamSource.cs:175]: I think this change came from changing the CRLF file back to LF, I'll re-add this space

File: Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (homuler) [reply Assets/Mediapipe/Samples/Common/Scripts/ImageSource/ImageSource.cs]: I don't think you can distinguish EOLs on GitHub, but you can usually do that with your editor by visualizing them.

And If you ran `git add .`, I recommend you run `git add -p .` instead.
You can notice when unintended changes are about to be staged.

**Comment by ROBYER1** [comment]
I'm revisiting this as I am also tackling device orientation Landscape left/right/portrait/portrait upside down

**Comment by ROBYER1** [comment]
I also looked into sorting the rotation of the onscreen annotation screen with the following code, however I think this may be something to do in another PR?

```  void Update()
        {
            if (_isInitialized == true && baseRotation != null)
            {
                if (Screen.orientation != _screenOrientation)
                {
                    _screenOrientation = Screen.orientation;

                    if (Screen.orientation == ScreenOrientation.Portrait)
                    {
                        screenTrans.eulerAngles = new Vector3(0, 0, baseRotation.z);
                        return;
                    }
                    else if (Screen.orientation == ScreenOrientation.PortraitUpsideDown)
                    {
                        screenTrans.eulerAngles = new Vector3(0, 0, baseRotation.z - 180);
                        return;
                    }
                    else if (Screen.orientation == ScreenOrientation.LandscapeLeft)
                    {
                        screenTrans.eulerAngles = new Vector3(0, 0, baseRotation.z + 90);
                        return;
                    }
                    else if (Screen.orientation == ScreenOrientation.LandscapeRight)
                    {
                        screenTrans.eulerAngles = new Vector3(0, 0, baseRotation.z - 90);
                        return;
                    }
                }
            }
        }

```

**Comment by homuler** [comment]
https://github.com/homuler/MediaPipeUnityPlugin/commit/f04ba48645c6668ff3ffa6d7cd51125b5add44e1 fixed it.

**Pull request closed without merging**